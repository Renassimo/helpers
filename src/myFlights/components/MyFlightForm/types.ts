import { MyFlightAttributes } from '@/myFlights/types';
import { Matcher } from '@/common/types/matchers';

export default interface MyFlightFormFieldProps {
  name: keyof MyFlightAttributes;
  label: string;
  options?: string[];
  matchers?: Matcher;
  type?: string;
  disabled?: boolean;
  isDate?: boolean;
}
