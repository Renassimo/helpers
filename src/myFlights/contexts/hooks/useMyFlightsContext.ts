import { useContext } from 'react';

import MyFlightsContext from '@/myFlights/contexts/MyFlightsContext';

const useMyFlightsContext = () => useContext(MyFlightsContext);

export default useMyFlightsContext;
