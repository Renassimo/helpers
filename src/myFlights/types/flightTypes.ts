export interface FlightAttributes {
  title: string | null;
  date: string | null;
  airline: string | null;
  flightNumber: string | null;
  origin: string | null;
  destination: string | null;
  manufacturer: string | null;
  model: string | null;
  registration: string | null;
  cn: string | null;
  firstFlight: string | null;
  airplaneName: string | null;
  originName: string | null;
  destinationName: string | null;
  seatNumber: string | null;
  altAirline: string | null;
  altFlightNumber: string | null;
  planespottersUrl: string | null;
  distance: number | null;
  age: string | null;
  url: string;
  photoUrl: string | null;
  number: number;
}

export interface FlightData {
  id: string;
  attributes: FlightAttributes;
}

export interface FlightsState {
  [id: string]: FlightData;
}

export interface FlightsContextData {
  // Flights data
  flights: FlightsState;
  flightsList: FlightData[];
  updateFlight: (flight: FlightData) => void;
}
