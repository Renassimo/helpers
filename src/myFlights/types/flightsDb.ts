import {
  NotionSelectProperty,
  NotionDBSelectProperties,
} from '@/common/types/notion';

export interface FlightsDBProperties extends NotionDBSelectProperties {
  Origin: NotionSelectProperty;
  Destination: NotionSelectProperty;
  Airline: NotionSelectProperty;
  Manufacturer: NotionSelectProperty;
  'Alt airline': NotionSelectProperty;
  Model: NotionSelectProperty;
}

export interface FlightsDB {
  object: 'database';
  id: string;
  properties: FlightsDBProperties;
}
