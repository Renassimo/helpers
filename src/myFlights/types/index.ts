export * from './myFlightTypes';
export * from './myFlightsDb';
export * from './props';
