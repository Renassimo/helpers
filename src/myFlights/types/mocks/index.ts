export * from './flights';
export * from './flightsDb';
