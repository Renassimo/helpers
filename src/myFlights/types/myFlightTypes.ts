export interface MyFlightAttributes {
  title: string | null;
  date: string | null;
  airline: string | null;
  flightNumber: string | null;
  origin: string | null;
  destination: string | null;
  manufacturer: string | null;
  model: string | null;
  registration: string | null;
  cn: string | null;
  firstFlight: string | null;
  airplaneName: string | null;
  originName: string | null;
  destinationName: string | null;
  seatNumber: string | null;
  altAirline: string | null;
  altFlightNumber: string | null;
  planespottersUrl: string | null;
  distance: number | null;
  age: string | null;
  url: string;
  photoUrl: string | null;
  number: number;
}

export interface MyFlightData {
  id: string;
  attributes: MyFlightAttributes;
}

export interface MyFlightsState {
  [id: string]: MyFlightData;
}

export interface FlightsContextData {
  // Flights data
  flights: MyFlightsState;
  flightsList: MyFlightData[];
  updateFlight: (flight: MyFlightData) => void;
}
