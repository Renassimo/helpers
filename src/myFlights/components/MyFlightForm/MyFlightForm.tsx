import Box from '@mui/material/Box';

import AircraftForm from '@/myFlights/components/AircraftForm';
import AirportForm from '@/myFlights/components/AirportForm';
import FlightForm from '@/myFlights/components/FlightForm';

import useMyFlightsContext from '@/myFlights/contexts/hooks/useMyFlightsContext';

const MyFlightForm = () => {
  const {
    flightsResult: { chosenFlight },
  } = useMyFlightsContext();

  const { attributes } = chosenFlight ?? {};
  const { origin, destination, registration } = attributes ?? {};

  return (
    <Box>
      <FlightForm />
      {!origin && <AirportForm title="Origin" />}
      {!destination && <AirportForm title="Destination" />}
      <AircraftForm registration={registration} />
    </Box>
  );
};

export default MyFlightForm;
