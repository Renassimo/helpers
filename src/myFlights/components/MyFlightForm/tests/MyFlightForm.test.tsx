import renderWithTheme from '@/common/tests/helpers/renderWithTheme';
import { Avia } from '@/avia/types/avia';

import useMyFlightsContext from '@/myFlights/contexts/hooks/useMyFlightsContext';
import useInputValue from '@/common/hooks/useInputValue';

import DateInput from '@/common/components/DateInput';
import FreeAutoComplete from '@/common/components/FreeAutoComplete';
import ClearableInput from '@/common/components/ClearableInput';

import MockedDateInput from '@/common/components/DateInput/mocks';
import MockedFreeAutoComplete from '@/common/components/FreeAutoComplete/mocks';
import MockedClearableInput from '@/common/components/ClearableInput/mocks';

import MyFlightForm from '../MyFlightForm';

jest.mock('@/myFlights/contexts/hooks/useMyFlightsContext');
jest.mock('@/common/hooks/useInputValue');
jest.mock('@/common/components/DateInput');
jest.mock('@/common/components/FreeAutoComplete');
jest.mock('@/common/components/ClearableInput');

describe('MyFlightForm', () => {
  const options: Avia.Options = {
    airlines: ['Airline 1', 'Airline 2'],
    airports: ['Airport 3', 'Airport 4'],
    manufacturers: ['Manufacturer 5', 'Manufacturers 6'],
    models: ['Model 7', 'Model 8'],
  };
  const matchers: Avia.Matchers = {
    airlines: { 'Airline 1': 'Airline 2' },
    airports: { 'Airport 3': 'Airport 4' },
    manufacturers: { 'Manufacturer 5': 'Manufacturers 6' },
    models: { 'Model 7': 'Model 8' },
  };

  beforeEach(() => {
    (DateInput as unknown as jest.Mock).mockImplementation(MockedDateInput);
    (FreeAutoComplete as unknown as jest.Mock).mockImplementation(
      MockedFreeAutoComplete
    );
    (ClearableInput as unknown as jest.Mock).mockImplementation(
      MockedClearableInput
    );
  });

  test('renders successfully with empty values', () => {
    // Arange
    (useMyFlightsContext as unknown as jest.Mock).mockImplementation(
      jest.fn(() => ({
        options,
        matchers,
        loadedValues: {},
      }))
    );
    (useInputValue as unknown as jest.Mock).mockImplementation(
      jest.fn(() => [{}, jest.fn()])
    );
    // Act
    const { baseElement } = renderWithTheme(<MyFlightForm />);
    // Assert
    expect(baseElement).toMatchSnapshot();
  });

  describe('when loaded values passed', () => {
    test('renders successfully with empty values', () => {
      // Arange
      const loadedValuesInState = {
        date: 'loadedValues.date',
        flightNumber: 'loadedValues.flightNumber',
        registration: 'loadedValues.registration',
        cn: 'loadedValues.cn',
        firstFlight: 'loadedValues.firstFlight',
        airplaneName: 'loadedValues.airplaneName',
        originName: 'loadedValues.originName',
        destinationName: 'loadedValues.destinationName',
        seatNumber: 'loadedValues.seatNumber',
        altAirline: 'loadedValues.altAirline',
        altFlightNumber: 'loadedValues.altFlightNumber',
        planespottersUrl: 'loadedValues.planespottersUrl',
        distance: 'loadedValues.distance',
        age: 'loadedValues.age',
        photoUrl: 'loadedValues.photoUrl',
      };
      (useMyFlightsContext as unknown as jest.Mock).mockImplementation(
        jest.fn(() => ({
          options,
          matchers,
          loadedValues: {
            ...loadedValuesInState,
            model: 'loadedValues.model',
            manufacturer: 'loadedValues.manufacturer',
            destination: 'loadedValues.destination',
            origin: 'loadedValues.origin',
            airline: 'loadedValues.airline',
          },
        }))
      );
      let setStateResult: any = {};
      const mockedSetState = jest.fn((fn) => {
        setStateResult = fn({});
      });
      (useInputValue as unknown as jest.Mock).mockImplementation(
        jest.fn(() => [{}, mockedSetState])
      );
      // Act
      const { baseElement } = renderWithTheme(<MyFlightForm />);
      // Assert
      expect(baseElement).toMatchSnapshot();
      expect(setStateResult).toEqual(loadedValuesInState);
    });
  });
});
