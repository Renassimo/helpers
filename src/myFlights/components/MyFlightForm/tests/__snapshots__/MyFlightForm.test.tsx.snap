// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyFlightForm renders successfully with empty values 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedDateInput - 
           - 
           - 
          Date
           - 
           -
           
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          Flight Number
           - 
           - 
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          Alt flight number
           - 
           - 
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          Seat number
           - 
           - 
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedFreeAutoComplete - 
           - 
          Airline
           - 
           -
           
          Airline 1 -
          Airline 2 -
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedFreeAutoComplete - 
           - 
          Alt airline
           - 
           -
           
          Airline 1 -
          Airline 2 -
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedFreeAutoComplete - 
           - 
          Origin
           - 
           -
           
          Airport 3 -
          Airport 4 -
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          Origin name
           - 
           - 
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedFreeAutoComplete - 
           - 
          Destination
           - 
           -
           
          Airport 3 -
          Airport 4 -
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          Destination name
           - 
           - 
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          Distance, km
           - 
          number
           - 
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedFreeAutoComplete - 
           - 
          Manufacturer
           - 
           -
           
          Manufacturer 5 -
          Manufacturers 6 -
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedFreeAutoComplete - 
           - 
          Model
           - 
           -
           
          Model 7 -
          Model 8 -
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          Registration
           - 
           - 
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          CN
           - 
           - 
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          Airplane name
           - 
           - 
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedDateInput - 
           - 
           - 
          First flight
           - 
           -
           
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          Age
           - 
           - 
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          Planespotters URL
           - 
          url
           - 
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          Photo URL
           - 
          url
           - 
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`MyFlightForm when loaded values passed renders successfully with empty values 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedDateInput - 
           - 
           - 
          Date
           - 
           -
           
          loadedValues.date
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          Flight Number
           - 
           - 
          loadedValues.flightNumber
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          Alt flight number
           - 
           - 
          loadedValues.altFlightNumber
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          Seat number
           - 
           - 
          loadedValues.seatNumber
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedFreeAutoComplete - 
           - 
          Airline
           - 
          loadedValues.airline
           -
           
          Airline 1 -
          Airline 2 -
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedFreeAutoComplete - 
           - 
          Alt airline
           - 
          loadedValues.altAirline
           -
           
          Airline 1 -
          Airline 2 -
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedFreeAutoComplete - 
           - 
          Origin
           - 
          loadedValues.origin
           -
           
          Airport 3 -
          Airport 4 -
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          Origin name
           - 
           - 
          loadedValues.originName
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedFreeAutoComplete - 
           - 
          Destination
           - 
          loadedValues.destination
           -
           
          Airport 3 -
          Airport 4 -
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          Destination name
           - 
           - 
          loadedValues.destinationName
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          Distance, km
           - 
          number
           - 
          loadedValues.distance
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedFreeAutoComplete - 
           - 
          Manufacturer
           - 
          loadedValues.manufacturer
           -
           
          Manufacturer 5 -
          Manufacturers 6 -
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedFreeAutoComplete - 
           - 
          Model
           - 
          loadedValues.model
           -
           
          Model 7 -
          Model 8 -
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          Registration
           - 
           - 
          loadedValues.registration
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          CN
           - 
           - 
          loadedValues.cn
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          Airplane name
           - 
           - 
          loadedValues.airplaneName
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedDateInput - 
           - 
           - 
          First flight
           - 
           -
           
          loadedValues.firstFlight
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          Age
           - 
           - 
          loadedValues.age
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          Planespotters URL
           - 
          url
           - 
          loadedValues.planespottersUrl
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
           - 
          Photo URL
           - 
          url
           - 
          loadedValues.photoUrl
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`MyFlightForm when state values passed renders successfully with empty values 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedDateInput - 
          state.date
           - 
           - 
          Date
           - 
           -
           
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
          state.flightNumber
           - 
          Flight Number
           - 
           - 
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
          state.altFlightNumber
           - 
          Alt flight number
           - 
           - 
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
          state.seatNumber
           - 
          Seat number
           - 
           - 
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedFreeAutoComplete - 
           - 
          Airline
           - 
           -
           
          Airline 1 -
          Airline 2 -
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedFreeAutoComplete - 
          state.altAirline
           - 
          Alt airline
           - 
           -
           
          Airline 1 -
          Airline 2 -
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedFreeAutoComplete - 
           - 
          Origin
           - 
           -
           
          Airport 3 -
          Airport 4 -
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
          state.originName
           - 
          Origin name
           - 
           - 
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedFreeAutoComplete - 
           - 
          Destination
           - 
           -
           
          Airport 3 -
          Airport 4 -
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
          state.destinationName
           - 
          Destination name
           - 
           - 
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
          state.distance
           - 
          Distance, km
           - 
          number
           - 
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedFreeAutoComplete - 
           - 
          Manufacturer
           - 
           -
           
          Manufacturer 5 -
          Manufacturers 6 -
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedFreeAutoComplete - 
           - 
          Model
           - 
           -
           
          Model 7 -
          Model 8 -
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
          state.registration
           - 
          Registration
           - 
           - 
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
          state.cn
           - 
          CN
           - 
           - 
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
          state.airplaneName
           - 
          Airplane name
           - 
           - 
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedDateInput - 
          state.firstFlight
           - 
           - 
          First flight
           - 
           -
           
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
          state.age
           - 
          Age
           - 
           - 
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
          state.planespottersUrl
           - 
          Planespotters URL
           - 
          url
           - 
        </div>
      </div>
      <div
        class="MuiBox-root css-164r41r"
      >
        <div>
          MockedClearableInput - 
          state.photoUrl
           - 
          Photo URL
           - 
          url
           - 
        </div>
      </div>
    </div>
  </div>
</body>
`;
